File: app/src/test/java/com/beautifulquran/domain/EnglishTypographyTest.kt
package com.beautifulquran.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class EnglishTypographyTest {
    @Test
    fun `adds stops at capitalized sentence boundaries and ayah end`() {
        assertEquals(
            listOf("And said", "a group", "(of) the Book.", "Believe", "in what."),
            EnglishTypography.punctuate(
                listOf("And said", "a group", "(of) the Book", "Believe", "in what"),
            ),
        )
    }

    @Test
    fun `proper and reverential capitals do not create false stops`() {
        assertEquals(
            listOf("Indeed", "Allah", "He gives it", "to whom", "He wills."),
            EnglishTypography.punctuate(
                listOf("Indeed", "Allah", "He gives it", "to whom", "He wills"),
            ),
        )
    }

    @Test
    fun `speech cues stay attached to capitalized content`() {
        assertEquals(
            listOf("your Lord.", "Say", "Indeed", "He said", "O my people."),
            EnglishTypography.punctuate(
                listOf("your Lord", "Say", "Indeed", "He said", "O my people"),
            ),
        )
    }
}

File: web/src/domain/__tests__/EnglishTypography.test.ts
import { describe, expect, it } from 'vitest'
import { punctuateEnglishGlosses } from '../EnglishTypography'

describe('punctuateEnglishGlosses', () => {
  it('adds a stop before a genuine capitalized sentence and at the ayah end', () => {
    expect(
      punctuateEnglishGlosses(['And said', 'a group', '(of) the Book', 'Believe', 'in what']),
    ).toEqual(['And said', 'a group', '(of) the Book.', 'Believe', 'in what.'])
  })

  it('does not treat proper and reverential capitals as sentence starts', () => {
    expect(punctuateEnglishGlosses(['Indeed', 'Allah', 'He gives it', 'to whom', 'He wills']))
      .toEqual(['Indeed', 'Allah', 'He gives it', 'to whom', 'He wills.'])
  })

  it('does not insert a stop between a speech cue and its capitalized content', () => {
    expect(punctuateEnglishGlosses(['your Lord', 'Say', 'Indeed', 'He said', 'O my people']))
      .toEqual(['your Lord.', 'Say', 'Indeed', 'He said', 'O my people.'])
  })

  it('preserves existing terminal punctuation', () => {
    expect(punctuateEnglishGlosses(['Why?', 'Then', 'listen!']))
      .toEqual(['Why?', 'Then', 'listen!'])
  })
})
